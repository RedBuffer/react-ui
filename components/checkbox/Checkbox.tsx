import * as React from 'react';
import classnames from 'classnames';

interface PropTypes
  extends Omit<
    React.DetailedHTMLProps<
      React.InputHTMLAttributes<HTMLInputElement>,
      HTMLInputElement
    >,
    'size'
  > {
  children: React.ReactNode | string;

  disabled?: boolean;

  size?: 'sm' | 'md' | 'lg';
}

const Component = ({
  children,
  className,
  disabled = false,
  size = 'md',
  ...restProps
}: PropTypes): React.ReactElement => {
  let padding: number;
  let textPaddingLeft: number;
  let width: number;

  if (size === 'sm') {
    padding = 1;
    width = 5;
    textPaddingLeft = 6;
  } else if (size === 'lg') {
    padding = 3;
    width = 8;
    textPaddingLeft = 10;
  } else {
    padding = 2;
    width = 6;
    textPaddingLeft = 8;
  }

  return (
    <label
      className={classnames(
        'm-2',
        'block',
        'relative',
        'select-none',
        `text-${size}`,
        'cursor-pointer',
        `pl-${textPaddingLeft}`,
        'react-ui-checkbox-label',
        className,
      )}
    >
      {children}
      <input
        className={classnames(
          'h-0',
          'w-0',
          'absolute',
          'opacity-0',
          'react-ui-checkbox',
        )}
        type="checkbox"
        disabled={disabled}
        {...restProps}
      />
      <span
        className={classnames(
          'top-0',
          'left-0',
          'absolute',
          'bg-gray-400',
          'rounded',
          `h-${width}`,
          `w-${width}`,
          `p-${padding}`,
          `checkmark-${size}`,
          'react-ui-checkbox-checkmark',
        )}
      />
    </label>
  );
};

export default Component;
