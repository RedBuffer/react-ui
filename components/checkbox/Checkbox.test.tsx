import * as React from 'react';
import { mount } from 'enzyme';

import Checkbox from './Checkbox';

describe('Checkbox', () => {
  let handleChange;
  let wrapper;

  beforeEach(() => {
    handleChange = jest.fn((e) => e.target.value);
    wrapper = mount(
      <Checkbox id={'checkbox'} onChange={handleChange}>
        checkbox
      </Checkbox>,
    );
  });

  it('should render correctly', () => {
    const component = wrapper;

    expect(component).toMatchSnapshot();
  });

  it('click on checkbox', () => {
    wrapper
      .find('#checkbox')
      .hostNodes()
      .simulate('change', { target: { checked: true } });

    expect(handleChange.mock.calls.length).toEqual(1);
    expect(handleChange.mock.calls[0][0].target.checked).toEqual(true);
  });

  it('click on disabled checkbox', () => {
    wrapper.setProps({ disabled: true });
    wrapper.find('#checkbox').hostNodes().simulate('click');

    expect(handleChange.mock.calls.length).toEqual(0);
  });
});
