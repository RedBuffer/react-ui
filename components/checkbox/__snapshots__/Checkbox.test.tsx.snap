// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render correctly with menu 1`] = `
<Component
  id="checkbox"
  onChange={[MockFunction]}
>
  <label
    className="m-2 block relative select-none text-md cursor-pointer pl-8 react-ui-checkbox-label"
  >
    checkbox
    <input
      className="h-0 w-0 absolute opacity-0 react-ui-checkbox"
      disabled={false}
      id="checkbox"
      onChange={[MockFunction]}
      type="checkbox"
    />
    <span
      className="top-0 left-0 absolute bg-gray-400 rounded h-6 w-6 p-2 checkmark-md react-ui-checkbox-checkmark"
    />
  </label>
</Component>
`;
